src/metabase/legacy_mbql/util.cljc

```diff
@@ -849,6 +849,16 @@
            [:field (id :guard integer?) opts]
            [id (:source-field opts)]))))
 
+(defn pred-matches-form?
+  "Check if `form` or any of its children forms match `pred`. Absolute most predicates will not match (this function is
+  used for validation), so calling this function before `matching-locations` is more efficient."
+  [form pred]
+  (cond
+    (pred form)        true
+    (map? form)        (reduce-kv (fn [b _ v] (or b (pred-matches-form? v pred))) false form)
+    (sequential? form) (reduce (fn [b x] (or b (pred-matches-form? x pred))) false form)
```

Comment:
Is it not worth using `reduced` for the map case and `some` for the sequential one?

Reply:
Following the assumption that the validation in the normal scenario is not going to fail (hence, pred will not match anywhere), using reduced/unreduced is unnecessary.

I'm not using `some` because it's using the inefficient seq-based iteration. I haven't yet added an optimized `some` to `metabase.util.performance` yet because it wasn't needed so far. If you think it's better to add it now and use here, then I will do that.

Reply:
No I don't think this warrants adding that helper. See your point about short-circuiting, happy as is.